### ISSUE TITLE
Memory usage

### ISSUE THREAD

**Issue by haraldwalnum** [ibpsa/project1-boptest#240]
When we have tested with longer simulation periods (e.g. 1 year), we have had some issues with the docker container crashing with OME errors (exit code 137).
When running docker on windows (through WSL) you need to allocate how much memory docker can use, so for now we have been able to solve this by increasing the memory allocation, but even with relatively simple emulators we need about 8GB of memory.
 
To investigate this further I did a test with the bestest_hydronic case. I run it for 180 days, with a control step of 1 hour. 
While running the simulation, I logged the memory allocation of the docker container. I have attached the output in a csv file and also a plot of the results.
The case is run three times. No control is applied, but the control step is still 1 hour, so the emulator is stopped and started again every hour (requests.post('{0}/advance'.format(url), data = {}).json())
After each case, results and KPIs are requested:
results = requests.get('{0}/results'.format(url)).json()
kpi = requests.get('{0}/kpi'.format(url)).json()
 
In the attached data, you can recognise a spike in the memory usage after each simulation. This corresponds to the get.result request. After that you can see 7 slightly smaller spikes, which corresponds to the get.kpi request.

![bestest_hydronic_memory](https://user-images.githubusercontent.com/53710940/95023736-ea363880-067e-11eb-9dbd-5ae4a2824cd0.png)
[bestest_hydronic_memory.txt](https://github.com/ibpsa/project1-boptest/files/5324037/bestest_hydronic_memory.txt)

 
My main concern is that this is still a relatively simple emulator, with few outputs, and that this will be a bigger issue with more complicated emulators.

This issue is probably linked with issue #124 and might be merged.

**Comment by EttoreZ** [comment]
@haraldwalnum I developed a  custom testcase which is more complex and I also have memory issues for the specific run. Furthermore if I run several simulations in sequence (changing control strategy), without closing the docker container, the memory usage rises easily up to 16 GB.

**Comment by dhblum** [comment]
I've done a little studying on this.  I repeated Harald's test with bestest_hydronic (180 days and step size of 1 hour) on Ubuntu 16.04.  I add a 60 second pause between calling ``/results`` (first spike at end of test) and calling ``/kpis`` (second spike at end of test).  I get:

![mem_baseline](https://user-images.githubusercontent.com/19480999/97629307-500aab80-1a04-11eb-95da-f28bf79a354e.png)

with KPIs:

```
KPI RESULTS 
-----------
ener_tot: 2844.72560261 kWh
emis_tot: 514.62838243 kg CO2
idis_tot: 0.0 ppmh
tdis_tot: 2176.55704909 Kh
time_rat: 1.24860027924e-06 
cost_tot: 57.5092068885 euro or $
```

Using the three following simulation options:

1. ``options['filter'] = self.output_names + self.input_names`` - stores only BOPTEST inputs and output variables in pyfmi result file.
2. ``options['CVode_options']['store_event_points'] = False`` - does not store data points at events.
3. ``self.options['ncp'] = int((end_time-start_time)/30)`` - sets sampling interval of results to 30s.

I get the following profile:

![mem_all3](https://user-images.githubusercontent.com/19480999/97629647-c3142200-1a04-11eb-9d9c-998e3b79c786.png)

with KPIs:

```
KPI RESULTS 
-----------
ener_tot: 2844.2468966 kWh
emis_tot: 514.541688833 kg CO2
idis_tot: 0.0 ppmh
tdis_tot: 2177.17333648 Kh
time_rat: 1.31278338448e-06 
cost_tot: 57.4997512773 euro or $
```

By far the largest impact is 3., since 30 seconds is typically longer than the solver steps, so less data is being stored.  As can be seen, there is minimal impact on KPI calculation, with improved performance in terms of memory requirements and computational speed of the simulation (since not as much data is being stored after each communication step).  For both, it's about a 4x improvement.

**Comment by haraldwalnum** [comment]
Thanks @dhblum, I think your solution looks good. My only concern is that 2 and 3 could have more impact on the KPIs for other testcases. I think especially about cases where there are some internal on-off controllers (e.g. thermostats), that turn on or off several times within a control step. In a worst case scenario, one could have a on-off controller that turns on and off with the same frequency as the output. Then you would measure it as always on or always off. This partly brings us back to the discussions we had with @JavierArroyoBastida in #213

**Comment by dhblum** [comment]
Good point.  We could set up a test.  Perhaps not 2 then.

**Comment by dhblum** [comment]
In thinking about this further, two other options are:

1. Add to the ``/results`` API an argument for variable name.  This is akin to how a typical database would be queried in a real building and limit the

**Comment by haraldwalnum** [comment]
I don't fully see the benefit of option 1. If I understand correctly we would still get the peaks.

However, what at about option 2, in combination with option 1 above, and then after calculation of the KPIs the results could be resampled to e.g. 30 seconds before they are added to a cumulative record. The full resolution result record can be overwritten for each control step.

**Comment by haraldwalnum** [comment]
One issue I see with my latest suggestion is if the results from Modelica have a longer time step than the specified resampling frequency

**Comment by haraldwalnum** [comment]
I made a quick testcase based on the BESTESTHydronic, by switching the radiator with a set point controlled electric heater.
To force rapid switching on/off, I set the power to 5 kW and the hysteresis to 0.5 K. The result is that the heater is on for about 30s before it switches off.

I gained the following KPIs after running for 1 week:
Old setup: 
KPIs are:
{   'cost_tot': 12.991883255760303,
    'emis_tot': 40.55410287312094,
    'ener_tot': 242.83893936000564,
    'idis_tot': 0,
    'tdis_tot': 0.6750476592545912,
    'time_rat': 3.4711183267518513e-06}

New setup:
KPIs are:
{   'cost_tot': 12.930291594225,
    'emis_tot': 40.36184478944999,
    'ener_tot': 241.68769335,
    'idis_tot': 0,
    'tdis_tot': 0.6707550774612209,
    'time_rat': 3.523545826742511e-06}

The difference in ener_tot is about 0.5%. It is not a lot, but probably also a bit random.
This case is a bit special, but no completely unrealistic. I log when all my heaters in my own home turn on and off, and today the I had several periods where heaters where on for about 1-2 minutes before they switch of.
However, I don't think any of the planned emulators have similar controls, so I don't know how important it is.

I still like the idea of integrating the power signals within the emulator, but I understand that many don't agree with me :).

**Comment by dhblum** [comment]
Integrating the power signals in the emulator also requires feeding price and emission data into the emulator during runtime.  Also to be consistent with thermal comfort and IAQ calculations, also feeding in the comfort range data during runtime.

Thanks for running the experiment. I agree, the error is dependent on the test case and controller, so will be hard to characterize.

> I don't fully see the benefit of option 1. If I understand correctly we would still get the peaks.

The peaks are from having to load all of the trajectory data for every variable at one time to pass through the rest api.  My suggestion was to add an argument to the ``/results`` api call to only retrieve data for one variable, much like a typical database works.  This greatly reduces the peak.  See below results for a test implementation where I only retrieve one variable in ``/results``.  The first peak is basically eliminated:

![mem_singleresult](https://user-images.githubusercontent.com/19480999/100150082-1c893880-2e6d-11eb-9fe9-4afef646d646.png)

In further exploration, the high memory accrued over simulation run time and computational overhead at each step can be improved by storing the variable trajectories more efficiently in memory.  In the case shown below, as np arrays instead of python lists (note this was implemented separately from the single variable strategy described above, so we still see the two peaks).  Note the smaller increase in memory from the start to the point where results are called, as well as a the faster simulation time.

![mem_nparray](https://user-images.githubusercontent.com/19480999/100150296-683be200-2e6d-11eb-8f5d-e4609b4e6043.png)

**Comment by JavierArroyoBastida** [comment]
I've been thinking about this further. Regarding KPI calculation I see two possibilities:
1. Integration within the emulator, as proposed in https://github.com/ibpsa/project1-boptest/issues/213#issuecomment-649995162. I must say that the more I think about it the more appealing this is. However, I still have few concerns about it. Following the nomenclature used in https://github.com/ibpsa/project1-boptest/issues/213#issuecomment-649995162, I 
don't think that it is the Pavg time series the one that should be multiplied by price or emission data, note that these variables are provided in units of EUR/kWh and kgCO2-e/kWh, respectively, so every time interval these should be multiplied by the energy increments: (Q_n-1 - Q_n), and not by average power. Python would only require to make a summation over these increments by their price/emission factors at each interval. This would indeed prevent having to feed the emulator with external data which makes this implementation quite smooth. One special case I see that is a bit more complex is the case of discomfort, since that one needs to be integrated only when outside a range, making it a bit more challenging. I think this can still be resolved using a conditional statement in Python that uses "positive" increments of the integrated temperature when temperature is above upper bound and "negative" increments of the integrated temperature when temperature is below upper bound. Let T be temperature and intT the integrated temperature signal. The pseudocode would be:
```
upper_discomfort =  0
lower discomfort = 0
for n=1,2,...,N:
    if T>T_upper_bound:
        upper_discomfort += (intT_n-1 - intT_n) 
    if T<T_lower_bound:
        lower_discomfort += -(intT_n-1 - intT_n) 

total_discomfort = upper_discomfort + lower_discomfort
```
Which can be much more efficiently implemented using pandas or numpy operators, but I'm using a for loop here for simplicity. 

2. Calculate core KPIs after each control step and add to cumulative record as proposed in https://github.com/ibpsa/project1-boptest/issues/240#issuecomment-725548420. This implementation may be easier to implement. 

I'm probably inclined for 1 right now...

**Comment by SenHuang19** [comment]
Based on the figures, I suspect there  is some left over memory that could accumulate over the course. I am not sure how Python or Pyfmi collect garbage  and remove all allocations and I wonder if we need to do the load_fmu to release all the memories. I will do some tests on the large office emulator and upload the results here.

**Comment by dhblum** [comment]
Indeed @JavierArroyoBastida 's new implementation of KPI calculation after each control step in https://github.com/ibpsa/project1-boptest/pull/265 reduces the memory spike seen in the previous figures (the second spike).  It also does not seem to add noticeable computation time at each step, since the time to complete the same test example still matches what was seen after the [numpy array storage improvement](https://github.com/ibpsa/project1-boptest/issues/240#issuecomment-733230224).

![mem](https://user-images.githubusercontent.com/19480999/102552503-55dd4e80-408f-11eb-8c2a-8efe67fc7202.png)

KPIs reported at the end are the same:

```
KPI RESULTS 
-----------
ener_tot: 2844.72560261 kWh
emis_tot: 514.62838243 KgCO2
idis_tot: 0.0 ppmh
tdis_tot: 2176.55704909 Kh
time_rat: 1.21805243494e-06 None
cost_tot: 57.5092068885 Euro or $
```

**Comment by haraldwalnum** [comment]
Looks like a good solution. However, i am still concerned that the memory usage is high. This is still a simple emulator... 
Since the KPIs now are calculated for each control step, could the resolution in the stored results be lower? It should be possible to resample the results that are stored for each control step, after the KPI calculation. I guess this should reduce the spike when requesting the results. 
A challenge might be that the data is not stored with an equidistant time step. However, this should be solvable. 

Not completely leaving the integration option:
- My suggestion with the discomfort KPIs (both temperature and CO2) would be to use the integrated value to calculate the average within each sample, and then use the average to calculate the KPIs in the same way that it is done now. The comfort parameters are normally quite slow, so if the sampling rate is not to low, this should be accurate enough? 
- For the implementation: If an integrator block is added to the "read" block in Modelica, and the "read" block outputs both the sampled value and the integrated value, it should only be necessary for the emulator to update the "read" blocks, without any further modifications.

**Comment by dhblum** [comment]
Thanks for the continued discussion @haraldwalnum. 

The spike in memory usage when using ``/results`` is part of the reason for my proposal in https://github.com/ibpsa/project1-boptest/issues/267, in addition to other points I make about how typical databases in BAS systems work anyway.  I demonstrated in https://github.com/ibpsa/project1-boptest/issues/240#issuecomment-733230224 how returning only one variable basically removes the spike due to the ``/results`` call.  But with my thoughts on how databases work, I think the change to the ``/results`` api is a good one anyway.

I'm not clear on what challenge you're referring to when you say "A challenge might be that the data is not stored with an equidistant time step."  But your note on storing resampled results is valid I think, since usually sensors don't read and data is not recorded at a "solver interval," but a regular resolution instead (except for data records that rely on change-of-value schemes rather than time-based schemes...).  The question is what that interval should be for each measurement in boptest.

On the integration within the emulator option, I'm still thinking about what the advantage is.  Here are some current thoughts.  With the implementation proposed in #265, the whole history of results no longer needs to be stored at the solver timestep.  That frees us to do long-term storage at any interval (prompting a discussion on what that interval should be that is no longer linked to the need to do the KPI integration).  And so, moving the integration into the emulators may have the advantage of less processing overhead at each timestep if Modelica/pyfmi can integrate faster than Python.  But that adds the averaging processing needed into Python you suggest, so not sure the tradeoff.  Would an advantage be if the number of communication points can be directly set in pyfmi (instead of taking the solver steps and resampling in Python) to reduce the size of the results file created during each simulation step?  Depending on the size of the emulator and length of the simulation step, this could have an impact on total simulation time.  See for instance my testing in https://github.com/ibpsa/project1-boptest/issues/240#issuecomment-719009346.

**Comment by dhblum** [comment]
Just an update, this is now the profile of current master (with https://github.com/ibpsa/project1-boptest/issues/258, https://github.com/ibpsa/project1-boptest/issues/264, and https://github.com/ibpsa/project1-boptest/issues/267 merged):

![mem](https://user-images.githubusercontent.com/19480999/107097059-18b55300-67da-11eb-94b5-10ab5834be55.png)

- Simulation accumulates about 18 minutes and 1.5 GB memory
- After 1 min sleep, results for the whole simulation time are retrieved in a for loop over all available input and measurement points (seen starting at 20 min)
- After 1 min sleep, KPIs are retrieved with virtually no memory usage or time.  Output is:
```
KPI RESULTS 
-----------
ener_tot: 2844.72560261 kWh
emis_tot: 514.62838243 KgCO2
idis_tot: 0.0 ppmh
tdis_tot: 2176.55704909 Kh
time_rat: 1.36661831927e-06 None
cost_tot: 57.5092068885 Euro or $
```

Looking better than where we started.

**Comment by dhblum** [comment]
Closed by #324 via #302.
